### ISSUE TITLE
Allow configuration of client with custom MantaApacheHttpClientContext

### ISSUE THREAD

**Issue by dkocher** [joyent/java-manta#310]
Allow to configure `MantaClient` with a custom context. We want to use our own configured `HttpClient` instance.

**Comment by tjcelaya** [comment]
Construction of the internal `HttpClient` is quite involved right now since the library has been designed to hide the http internals as much as possible. Would you expect to pass us a fully-configured `HttpClient` or perhaps a `HttpClientConnectionManager` like `PoolingHttpClientConnectionManager`?

**Comment by dkocher** [comment]
Depends on #307.

**Comment by dkocher** [comment]
It would probably suffice if setting the `MantaConnectionFactory` is possible which would allow to override `createBuilder`.

**Comment by dekobon** [comment]
Even if the library was to be unshaded, I would be hesitant to pass in an instance `HttpClient`. Here's a quick pass of the customization involved by default:

* Custom DNS resolver to handle multiple A records
* Interceptor for performing HTTP signature based authentication
* JMX monitoring of connection pool
* TCP settings: NoDelay, SoTimeout, SoKeepAlive
* Customization HTTP stream copy buffer size
* TLS/SSL cipher configuration
* Maximum connections settings
* Custom user-agent
* Custom request timeout
* Custom default headers
* Custom back off strategy
* Custom retry handler
* HTTP proxy settings are configurable

Half of these customization to a `HttpClient` instance are needed for the environment and the other half are optimizations. I could imagine us modifying the settings of an existing `HttpClient` instance so that it conforms to the above settings. However, that would mean that the instance would be totally unusable for anything outside of the Java Manta SDK.

**Comment by dkocher** [comment]
I would opt to allow to pass in a custom `HttpClientBuilder` that you can further configure with your custom interceptors required for authentication. All other options we want to set ourselves with settings from the user or native environment. In our case, proxy settings are read from system proxy settings and TLS is configured with trust managers reading certificates and trust settings from the Windows Store or Keychain on macOS respectively. We don't want the library to change our TLS trust settings.

**Comment by dekobon** [comment]
If we were going to accept a custom `HttpClientBuilder`, I think we would still need to do:

* Custom DNS resolver to handle multiple A records
* Interceptor for performing HTTP signature based authentication
* Custom default headers
* Custom back off strategy (or turn this off - 503 doesn't do a back off it just errors)
* Custom retry handler (or turn this off - no retires)

**Comment by dkocher** [comment]
I don't see how this issue is resolved.

**Comment by tjcelaya** [comment]
This was closed automatically because the issue number was included in another PR, apologies for the confusion.
